### PR TITLE
use GenMaps for trigger ACS

### LINKED ISSUE #7112: Use GenMap in DAML Trigger API
Currently, triggers do a bunch of linear searches because we cannot have maps indexed by contract ids and type-indexed maps. This should be changed to use maps internally for DAML-LF 1.dev.

This will require CPP for now.

### PR DESCRIPTION
**Pull request by S11001001**
Advances #7112 by introducing GenMaps in the triggers DAML library.

* [x] multiply tests by LF version where appropriate (integration: default, 1.dev)
* [x] revert to unqualified (no -1.dev) tests otherwise
* [x] changelog for new Ords, or none? none

### Test times

A few tests seem relevant, here are their timings locally should we consider multiplying them across LF versions so as to exercise the separate CPP-controlled paths:

```
//triggers/service:tests                                                 PASSED in 135.9s
//triggers/tests:trigger-integration-tests                               PASSED in 51.3s
//triggers/tests/scenarios:test-rule                                     PASSED in 17.4s
```

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/daml/BUILD.bazel [hunk 1]
```diff
@@ -39,21 +39,23 @@ dependencies:
...
 EOF
-          $(location //compiler/damlc) build --project-root=$$TMP_DIR -o $$PWD/$@
+          $(location //compiler/damlc) build --project-root $$TMP_DIR \
+              --cpp $$PWD/$(location @stackage-exe//hpp) \
+              -o $$PWD/$@
           rm -rf $$TMP_DIR
         """.format(
             build_options = str([
-                "--ghc-option",
-                "-Werror",
```
Comment (S11001001) [inline triggers/daml/BUILD.bazel:47]: Without which

```
hpp: Couldn't open input file: -Werror
CallStack (from HasCallStack):
  error, called at src/Hpp/CmdLine.hs:103:22 in hpp-0.6.1:Hpp.CmdLine
```

Comment (cocreature) [reply triggers/daml/BUILD.bazel:47]: Probably worth opening an issue for. Ideally, we would fix whatever is broken here and get back to the point where we can enable it.

File: triggers/daml/Daml/Trigger/Internal.daml [hunk 2]
```diff
@@ -30,7 +34,11 @@ import Daml.Trigger.LowLevel hiding (Trigger)
 
 -- This will change to a Map once we have proper maps in DAML-LF
 data ACS = ACS
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/Internal.daml:36]: Not for this PR but we could consider changing the list in the values to `Map AnyContractId AnyTemplate`. Then we can provide a fast lookup by contract id which is something you need fairly often.

File: triggers/daml/Daml/Trigger/Internal.daml [hunk 3]
```diff
@@ -45,15 +53,53 @@ newtype TriggerA a = TriggerA (State TriggerAState a)
 addCommands : Map CommandId [Command] -> Commands -> Map CommandId [Command]
 addCommands m (Commands cid cmds) = Map.insert cid cmds m
 
+#ifdef DAML_GENMAP && DAML_GENERIC_COMPARISON
+adjust : forall k v. Ord k => k -> (Optional v -> Optional v) -> GMap.Map k v -> GMap.Map k v
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/Internal.daml:57]: @sofiafaro-da perhaps something worth adding to `DA.Map`? What do you think? (not in this PR)

**Review by cocreature — APPROVED** [review]
Nice work, thank you very much!

**Comment by cocreature** [comment]
As for your questions:

1. If we can reasonably easily test the trigger integration tests against the default LF target and 1.dev, I think this is worth doing. I wouldn’t bother testing all LF versions and I wouldn’t bother doing this for the trigger service tests.
2. For the ones where we don’t test multiple LF versions (mostly thinking of the trigger service tests), keeping the default LF version seems very sensible.
3. I wouldn’t bother adding something here. We just filter out changes that only apply to 1.dev anyway when we write up the release notes.

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 3 above)
Comment (sofiafaro-da) [reply triggers/daml/Daml/Trigger/Internal.daml:57]: yeah, that looks like a good function to have in DA.Map :)

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 2 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/Internal.daml:38]: #7425

File: triggers/daml/Daml/Trigger/Internal.daml (hunk 3 above)
Comment (S11001001) [reply triggers/daml/Daml/Trigger/Internal.daml:59]: #7427

**Pull request merged**